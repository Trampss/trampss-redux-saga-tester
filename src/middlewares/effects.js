export default (mocks) => {
  const indexes = {}

  const initIndex = (type) => {
    if (indexes[type] === undefined) indexes[type] = 0
  }

  const getIndex = (type) => {
    initIndex(type)
    return indexes[type]
  }

  const incrementIndex = (type) => {
    initIndex(type)
    indexes[type] += 1
    return getIndex(type)
  }

  const getMockedValue = type => (params) => {
    const index = getIndex(type)
    incrementIndex(type)

    const mock = mocks && mocks[type] && mocks[type][index]
    if (!mock) return undefined

    return mock(params)
  }

  return (next) => {
    const { value } = next

    // no value -> leave
    if (!value) return next

    // not a saga effect -> leave
    if (!value['@@redux-saga/IO']) return next

    // read effects
    let type
    if (value.CALL) type = 'call'
    else if (value.SELECT) type = 'select'
    if (type) {
      return {
        raw: next,
        value: getMockedValue(type)(/* params */),
      }
    }

    // unmocked saga effect
    return next
  }
}
