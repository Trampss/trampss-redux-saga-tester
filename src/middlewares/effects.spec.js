/* eslint-env jest */
import effects from './effects'

describe('effects', () => {
  describe('unmodified next', () => {
    it('should not handle next with no value', () => {
      const next = { noValue: true }
      expect(effects(/* no mock */)(next)).toBe(next)
    })

    it('should not handle anything else than a redux-saga effect', () => {
      const next = { value: { noReduxSaga: true } }
      expect(effects(/* no mock */)(next)).toBe(next)
    })

    it('should not handle a reading effect', () => {
      const next = { value: { '@@redux-saga/IO': true, PUT: true } }
      expect(effects(/* no mock */)(next)).toBe(next)
    })
  })

  describe('modified next', () => {
    it('should return undefined value if there is no mock', () => {
      // CALL
      let next = { value: { '@@redux-saga/IO': true, CALL: true } }
      expect(effects(/* no mock */)(next)).toMatchSnapshot()

      // SELECT
      next = { value: { '@@redux-saga/IO': true, SELECT: true } }
      expect(effects(/* no mock */)(next)).toMatchSnapshot()
    })

    it('should return mocked informations', () => {
      const mocks = {
        call: [
          // first mock
          () => ({ first: 'mock' }),
          // second mock
          () => [{ second: 'mock' }, { should: 'be printed' }],
        ],
        select: [
          () => ({ select: 'mock too' }),
        ],
      }

      // CALL
      let next = { value: { '@@redux-saga/IO': true, CALL: true } }

      // first mock
      expect(effects(mocks)(next)).toMatchSnapshot()

      // first mock (because this is a new instance)
      const instance = effects(mocks)
      expect(instance(next)).toMatchSnapshot()

      // second mock (because same instance as before)
      expect(instance(next)).toMatchSnapshot()

      // SELECT
      next = { value: { '@@redux-saga/IO': true, SELECT: true } }
      expect(effects(mocks)(next)).toMatchSnapshot()
    })
  })
})
