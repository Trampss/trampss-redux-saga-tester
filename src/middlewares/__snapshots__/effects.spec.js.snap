// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effects modified next should return mocked informations 1`] = `
Object {
  "raw": Object {
    "value": Object {
      "@@redux-saga/IO": true,
      "CALL": true,
    },
  },
  "value": Object {
    "first": "mock",
  },
}
`;

exports[`effects modified next should return mocked informations 2`] = `
Object {
  "raw": Object {
    "value": Object {
      "@@redux-saga/IO": true,
      "CALL": true,
    },
  },
  "value": Object {
    "first": "mock",
  },
}
`;

exports[`effects modified next should return mocked informations 3`] = `
Object {
  "raw": Object {
    "value": Object {
      "@@redux-saga/IO": true,
      "CALL": true,
    },
  },
  "value": Array [
    Object {
      "second": "mock",
    },
    Object {
      "should": "be printed",
    },
  ],
}
`;

exports[`effects modified next should return mocked informations 4`] = `
Object {
  "raw": Object {
    "value": Object {
      "@@redux-saga/IO": true,
      "SELECT": true,
    },
  },
  "value": Object {
    "select": "mock too",
  },
}
`;

exports[`effects modified next should return undefined value if there is no mock 1`] = `
Object {
  "raw": Object {
    "value": Object {
      "@@redux-saga/IO": true,
      "CALL": true,
    },
  },
  "value": undefined,
}
`;

exports[`effects modified next should return undefined value if there is no mock 2`] = `
Object {
  "raw": Object {
    "value": Object {
      "@@redux-saga/IO": true,
      "SELECT": true,
    },
  },
  "value": undefined,
}
`;
